Agent restart fails due to interrupted system call on bpf node configuration file write

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

higher than v1.16.0 and lower than v1.17.0

### What happened?

Agent restart failed. All endpoints are in "not-ready" state after bpf compilation of bpf_lxc.c fails. The apparent reason is an interrupted system call on the node config file write:
```
time="2024-08-12T16:47:32Z" level=error msg="Unable to write node config header" error="failed to write node configuration file at /var/run/cilium/state/globals/node_config.h: interrupted system call" subsys=datapath-loader
```
Which leads to missing C macro definitions for the BPF compilation.

This could potentially be fixed by retrying system calls when interrupted?

### How can we reproduce the issue?

not easily reproducible

### Cilium Version

tip of `main` as on 13.8.2024
```
time="2024-08-12T16:47:30Z" level=info msg="Cilium 1.17.0-dev 93db3f9e 2024-08-12T18:27:02+02:00 go version go1.22.6 linux/amd64" subsys=daemon
```

### Kernel Version

n/a

### Kubernetes Version

n/a, happened in a runtime test without k8s

### Regression

_No response_

### Sysdump

[cilium-sysdumps (1).zip](https://github.com/user-attachments/files/16594848/cilium-sysdumps.1.zip)


### Relevant log output

```shell
time="2024-08-12T16:47:32Z" level=error msg="Unable to write node config header" error="failed to write node configuration file at /var/run/cilium/state/globals/node_config.h: interrupted system call" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=info msg="regenerating all endpoints" reason="Configuration or devices changed" subsys=endpoint-manager
time="2024-08-12T16:47:32Z" level=info msg="Reusing existing global key" key=Y29udGFpbmVyOmlkLmNsaWVudD07 subsys=allocator
time="2024-08-12T16:47:32Z" level=info msg="Reusing existing global key" key=Y29udGFpbmVyOmlkLnNlcnZlcj07 subsys=allocator
time="2024-08-12T16:47:32Z" level=error msg="Failed to compile bpf_host.o: exit status 1" compiler-pid=568 subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="In file included from /var/lib/cilium/bpf/bpf_host.c:29:" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="In file included from /var/lib/cilium/bpf/lib/common.h:1275:" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="In file included from /var/lib/cilium/bpf/lib/overloadable.h:10:" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="In file included from /var/lib/cilium/bpf/lib/overloadable_skb.h:8:" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="/var/lib/cilium/bpf/lib/clustermesh.h:18:23: error: use of undeclared identifier '__config_identity_length'" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="   18 |         __u32 identity_len = CONFIG(identity_length);" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="      |                              ^" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="/var/lib/cilium/bpf/lib/static_data.h:75:10: note: expanded from macro 'CONFIG'" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="   75 |         (typeof(__config_##name))out; \\" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="      |                 ^" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="<scratch space>:52:1: note: expanded from here" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="   52 | __config_identity_length" subsys=datapath-loader
time="2024-08-12T16:47:32Z" level=warning msg="      | ^" subsys=datapath-loader
```


### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/datapath/linux/config/config.go
+++ b/pkg/datapath/linux/config/config.go
@@ -305,7 +305,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		cDefinesMap["ENABLE_WIREGUARD"] = "1"
 		ifindex, err := link.GetIfIndex(wgtypes.IfaceName)
 		if err != nil {
-			return err
+			return fmt.Errorf("getting %s ifindex: %w", wgtypes.IfaceName, err)
 		}
 		cDefinesMap["WG_IFINDEX"] = fmt.Sprintf("%d", ifindex)
 
@@ -540,7 +540,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
 	macByIfIndexMacro, isL3DevMacro, err := devMacros(nativeDevices)
 	if err != nil {
-		return err
+		return fmt.Errorf("generating device macros: %w", err)
 	}
 	cDefinesMap["NATIVE_DEV_MAC_BY_IFINDEX(IFINDEX)"] = macByIfIndexMacro
 	cDefinesMap["IS_L3_DEV(ifindex)"] = isL3DevMacro
@@ -719,7 +719,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
 	vlanFilter, err := vlanFilterMacros(nativeDevices)
 	if err != nil {
-		return err
+		return fmt.Errorf("rendering vlan filter macros: %w", err)
 	}
 	cDefinesMap["VLAN_FILTER(ifindex, vlan_id)"] = vlanFilter
 
@@ -754,12 +754,12 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
 	ephemeralMin, err := getEphemeralPortRangeMin(h.sysctl)
 	if err != nil {
-		return err
+		return fmt.Errorf("getting ephemeral port range minimun: %w", err)
 	}
 	cDefinesMap["EPHEMERAL_MIN"] = fmt.Sprintf("%d", ephemeralMin)
 
 	if err := cDefinesMap.Merge(h.nodeExtraDefines); err != nil {
-		return err
+		return fmt.Errorf("merging extra node defines: %w", err)
 	}
 
 	for _, fn := range h.nodeExtraDefineFns {
@@ -769,7 +769,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		}
 
 		if err := cDefinesMap.Merge(defines); err != nil {
-			return err
+			return fmt.Errorf("merging extra node define func results: %w", err)
 		}
 	}
 
@@ -777,14 +777,14 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		if option.Config.IPv4Enabled() {
 			ipip4, err := netlink.LinkByName(defaults.IPIPv4Device)
 			if err != nil {
-				return err
+				return fmt.Errorf("looking up link %s: %w", defaults.IPIPv4Device, err)
 			}
 			cDefinesMap["ENCAP4_IFINDEX"] = fmt.Sprintf("%d", ipip4.Attrs().Index)
 		}
 		if option.Config.IPv6Enabled() {
 			ipip6, err := netlink.LinkByName(defaults.IPIPv6Device)
 			if err != nil {
-				return err
+				return fmt.Errorf("looking up link %s: %w", defaults.IPIPv6Device, err)
 			}
 			cDefinesMap["ENCAP6_IFINDEX"] = fmt.Sprintf("%d", ipip6.Attrs().Index)
 		}
@@ -870,7 +870,7 @@ func vlanFilterMacros(nativeDevices []*tables.Device) (string, error) {
 
 	links, err := netlink.LinkList()
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("listing network interfaces: %w", err)
 	}
 
 	for _, l := range links {

